File: packages/ui/db/entity/Route.ts
import { Boulder, Commentable, CoordinateOptional, Photoable } from "@/db";
import { cascadeManyToOne } from "@/db/cascadeOptions";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import { type IRoute } from "models";
import { EntitySchema } from "typeorm";
import CoordinateOptionalSchema from "./CoordinateOptional";

export type Route = IRoute & {
  coordinates?: CoordinateOptional;
  boulder?: Boulder;
  commentable?: Commentable;
  photoable?: Photoable;
};

const RouteSchema = new EntitySchema<Route>({
  name: "route",
  columns: {
    ...BaseColumnSchemaPart,
    name: {
      type: String,
    },
    length: {
      type: Number,
      nullable: true,
    },
    description: {
      type: String,
      nullable: true,
    },
    firstAscent: {
      type: String,
      nullable: true,
    },
    gradeRaw: {
      type: String,
      nullable: true,
    },
  },
  relations: {
    boulder: {
      type: "many-to-one",
      target: "boulder",
      ...cascadeManyToOne,
    },
    commentable: {
      type: "one-to-one",
      nullable: true,
      target: "commentable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
    photoable: {
      type: "one-to-one",
      nullable: true,
      target: "photoable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
  },
  embeddeds: {
    coordinates: {
      schema: CoordinateOptionalSchema,
      prefix: "coordinates_",
    },
  },
});

export default RouteSchema;

File: go-server/models/route.go
package models

type Route struct {
	ID          *int     `json:"id,omitzero"`
	Name        string   `json:"name"`
	Length      *int     `json:"length,omitzero"`
	Description *string  `json:"description,omitzero"`
	FirstAscent *string  `json:"firstAscent,omitzero"`
	GradeRaw    string   `json:"gradeRaw"`
	Grade       *Grade   `json:"grade,omitzero"`
	Boulder     *Boulder `json:"boulder,omitzero"`
}

func (r Route) IsZero() bool {
	return r.ID == nil && r.Name == "" && r.Length == nil && r.Description == nil && r.FirstAscent == nil && r.GradeRaw == "" && r.Grade == nil && r.Boulder == nil
}
